### PR TITLE
feat(settings): Toggle settings nav for smaller browser widths

### PR DESCRIPTION
**Pull request by jmanhart**
Adding in a hamburger menu for opening and closing the settings side panel for browser widths below 768px.

**Menu Open**
menu-open

**Menu Closed**
menu-closed

#sync-getsentry

### CODE REVIEW

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: Can we have it float over the content?

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: nit: would be nice to have space between this statment, since that matches what we typically do

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: You can make this `priority="link"` or something along those lines to get rid of the button styles.

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: Use the `space` helper here.

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: You can leave the spaces

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: Would be nice to see some animations in here too, shouldn't be hard to get this to slide in if we do something like animate it from `left: -100%` to `left: 0`.

**Comment by dashed** [comment]
👀 👀 👀

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: There's an emotion helper for this. https://github.com/getsentry/sentry/blob/7aa676f5f62592407a26d8beeebb03384aecf0cb/src/sentry/static/sentry/app/styles/animations.tsx#L25-L41

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: I think we need to get rid of the display if we're going to animate it

**Comment by jmanhart** [comment]
**At Rest**
Screen Shot 2019-12-06 at 5 32 14 PM

**Open Menu**
Screen Shot 2019-12-06 at 5 32 23 PM

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (jmanhart) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: I was having issues passing props to the keyframes helper. @dashed found this article highlighting the bug. 
https://github.com/styled-components/styled-components/issues/397

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (billyvg) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: Just a note here, [`setState` is async,](https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous) so it's possible that `this.state.navVisible` is not the most up to date value. When we update state that relies on previous state, it's better practice to use the callback version of `setState()`

```suggestion
                onClick={() => this.setState(state => ({navVisible: !state.navVisible}))}
```

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (billyvg) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: Is this something we should put in `theme.tsx`?

Comment (jmanhart) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: @billyvg It might fit better in the `animations.tsx`? Maybe I could use the `slideInLeft` from that file.

**Comment by jmanhart** [comment]
**Menu Open**
menu-closed-new

**Menu  Closed**
menu-open-new

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (billyvg) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: We should use `componentDidMount` as `WillMount` is deprecated

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (billyvg) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: Hmm is this a bootstrap class?

Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: Yeah I think it is. Would be open to a better solution here for sure.

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: bit of a nit, but we try and group our imports in order of:

1. 3rd party imports
2. app imports
3. relative imports

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: nit: I wish prettier would put a space between this function. Maybe we can eslint for this /cc @billyvg, what do you think?

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: Since the `router` needs to be passed now, we'll need to update the settingsLayout tests to take the router param

Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: Fixed in 619c649

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: Fixed in 619c649

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: Fixed in 2032ada

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: Fixed in 619c649

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: Were you talking about the `z-index` here @billyvg ?

**Review by evanpurkhiser — DISMISSED** [review]

**Comment by EvanPurkhiser** [comment]
I've taken over this PR. Going to dismiss my review and ask for a final review.

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (billyvg) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: Should we move this into theme?

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx
Comment (billyvg) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.jsx]: Yes `z-index`

**Comment by billyvg** [comment]
o.o what's the status of this? @EvanPurkhiser @jmanhart

**Comment by EvanPurkhiser** [comment]
I think I have a branch with some things rebased, but need to rebase again

**Comment by billyvg** [comment]
o.o

**Comment by EvanPurkhiser** [comment]
I know.. this needs revisited :(

**Comment by billyvg** [comment]
There are some views that don't look great, but better than existing views -- I didn't really see anything was broken

**Comment by matejminar** [comment]
I still need to address one issue with the current implementation - you can't scroll all the way to the bottom of the navigation on the iPhone.

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (matejminar) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:77]: Should we extract this to CSS class and toggle just the class?

Comment (markstory) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:77]: :+1: Toggling a class would be simpler and makes the restore process less fragile too.

File: tests/js/spec/views/settings/settingsIndex.spec.jsx
Comment (markstory) [inline tests/js/spec/views/settings/settingsIndex.spec.jsx:17]: Are there any functional tests we could add for the new behavior? Should we have an acceptance test with a visual snapshot that covers the new menu?

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:28]: I'd convert the propTypes to typescript :)

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:92]: You don't need `React.Fragment` here, because you are already wrapping the components using `SettingsColumn`😉

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:147]: I believe you don't need `&` here

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:167]: are you sure that these props are not being added to the dom?

Comment (matejminar) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:167]: They are in camelCase, we should be safe.

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (matejminar) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:147]: Should we be consistent with the rest of the codebase? Everywhere else we are using `&` in front of pseudo-selectors. It is working without it, but I think we should decide on one style and use it everywhere - maybe something to talk about on our TSC?

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (matejminar) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:92]: Good catch!

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (matejminar) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:34]: I need that constructor for other things anyway, but I can move the state declaration to a class property.

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (matejminar) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:27]: They are already in typescript props, but I also left proptypes, because this component is still used in a few JSX files.

File: tests/js/spec/views/settings/settingsIndex.spec.jsx
Comment (matejminar) [reply tests/js/spec/views/settings/settingsIndex.spec.jsx:17]: Added 👍

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:168]: I would check it anyway.. just in case... this week I used isClipped..and thought the same, but when I checked the dom..it was there 😉

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:148]: yes we can discuss it in the next TSC 😉

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:35]: just out of curiosity.. why did you need the constructor?

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:28]: oh yes.. I just saw it 👌

**Review by priscilawebdev — COMMENTED** [review]
.

**Comment by markstory** [comment]
Visual diffs for this look great. The mobile viewports for settings are actually visible now. :clap:

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:38]: I would move this to the line where you have `headerRef: React.RefObject<HTMLDivElement>;` you could write it as t`his.headerRef = React.createRef<HTMLDivElement>();`

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:41]: wouldn't be better to move it to the componentDidMount? it's recommended to avoid introducing any side-effects or subscriptions in the constructor

**Comment by priscilawebdev** [comment]
I tested it locally and it looks great  🚀

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (markstory) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:44]: Nit: If these comments were on the type then they would show up in autocomplete.

**Review by markstory — COMMENTED** [review]
Looked good to me locally.

File: src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/settings/components/settingsLayout.tsx:114]: wouldn't it be better to check if this function is available otherwise make it required?

**Review by markstory — APPROVED** [review]

**Pull request merged**
